incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -64,11 +64,24 @@ func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Resu
 	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/501
 	inst, err := r.getInstance(ctx, pnm, nm)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			// If the instance doesn't exist, return nil to prevent a retry.
+		return ctrl.Result{}, err
+	}
+
+	// If the instance is being deleted, remove the finalizer or cascading delete.
+	if !inst.DeletionTimestamp.IsZero() {
+		if len(inst.ObjectMeta.Finalizers) == 0 {
 			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, err
+		updateFinalizer := r.syncDeleting(log, inst)
+		if updateFinalizer {
+			r.writeInstance(ctx, log, inst)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// If the instance doesn't exist, return nil to prevent a retry.
```

Comment:
Let's say we never call syncDeleting because there were no finalizers. Do we need to do anything else here?

I think the comment should say something like "Someone enqueued this nonexisting instance by mistake, or else it's just been deleted. Either way, there's nothing we need to do." We shouldn't say things like "return nil to prevent a retry" because that just says _what_ the code is doing, and not _why_ the code is doing it. We should return nil because "there's no error, or we've already handled the error," not because "we don't want a retry."

Reply:
Update: we just discussed this in our 1:1. It's fine to leave it for this PR, but in the future we should handle the "the HNS has _been_ deleted here" case.